drivers/dma/dma_stm32_v3.c

```diff
@@ -0,0 +1,690 @@
+	if (ret < 0) {
+		return ret;
+	}
+
+	channel_config->busy = true;
+	channel_config->dma_callback = config->dma_callback;
+	channel_config->direction = config->channel_direction;
+	channel_config->user_data = config->user_data;
+	channel_config->src_size = config->source_data_size;
+	channel_config->dst_size = config->dest_data_size;
+	channel_config->cyclic = config->cyclic;
+
+	dma_stm32_disable_it(dma, channel);
+
+	if (config->block_count == 1 && !config->cyclic) {
+		ccr |= LL_DMA_LSM_1LINK_EXECUTION;
+	} else {
+		LOG_ERR("Only single block transfers are supported for now");
+		return -ENOTSUP;
+	}
+
+	/* TODO: support port specifier from configuration */
+	ccr |= ll_priority;
+	ccr |= LL_DMA_LINK_ALLOCATED_PORT0;
+
+	LL_DMA_ConfigControl(dma, channel, ccr);
+
+	hwdesc.channel_tr1 =
+		dest_inc_mode | dest_data_width_size | src_inc_mode | src_data_width_size;
+
+	LL_DMA_ConfigTransfer(dma, channel, hwdesc.channel_tr1);
+
+	LL_DMA_ConfigBurstLength(dma, channel, config->source_burst_length,
+				 config->dest_burst_length);
+
+	hwdesc.channel_tr2 = ll_direction;
+
+	if (config->complete_callback_en == LL_DMA_TCEM_BLK_TRANSFER) {
+		hwdesc.channel_tr2 |= LL_DMA_TCEM_BLK_TRANSFER;
+	} else {
+		hwdesc.channel_tr2 |= LL_DMA_TCEM_EACH_LLITEM_TRANSFER;
+	}
+
+	LL_DMA_ConfigChannelTransfer(dma, channel, hwdesc.channel_tr2);
+
+	if (ll_direction != LL_DMA_DIRECTION_MEMORY_TO_MEMORY) {
+		LL_DMA_SetPeriphRequest(dma, channel, config->dma_slot);
+	}
+
+	hwdesc.channel_br1 = config->head_block->block_size;
+
+	LL_DMA_SetBlkDataLength(dma, channel, hwdesc.channel_br1);
+
+	hwdesc.channel_sar = config->head_block->source_address;
+	hwdesc.channel_dar = config->head_block->dest_address;
+
+	LL_DMA_ConfigAddresses(dma, channel, hwdesc.channel_sar, hwdesc.channel_dar);
+
+	dma_stm32_enable_it(dma, channel);
+
+	return 0;
+}
+
+static int dma_stm32_reload(const struct device *dev, uint32_t id, uint32_t src, uint32_t dst,
+			    size_t size)
+{
+	const struct dma_stm32_config *dev_config;
+	struct dma_stm32_channel *channel_config;
+	uint32_t channel;
+	DMA_TypeDef *dma;
+
+	if (dev == NULL) {
+		LOG_ERR("Invalid device pointer");
+		return -EINVAL;
+	}
+
+	dev_config = dev->config;
+
+	if (id >= dev_config->max_channels) {
+		LOG_ERR("Invalid channel ID %d, max channels %d", id, dev_config->max_channels);
+		return -EINVAL;
+	}
+
+	channel_config = &dev_config->channels[id];
+	dma = dev_config->base;
+	channel = dma_stm32_id_to_channel(id);
+
+	if (dma_stm32_disable_channel(dma, channel) != 0) {
+		return -EBUSY;
+	}
+
+	LL_DMA_ConfigAddresses(dma, channel, src, dst);
+
+	LL_DMA_SetBlkDataLength(dma, channel, size);
+
+	channel_config->busy = true;
+
+	LL_DMA_EnableChannel(dma, channel);
+
+	return 0;
+}
+
+static int dma_stm32_start(const struct device *dev, uint32_t id)
+{
+	const struct dma_stm32_config *dev_config = dev->config;
+	struct dma_stm32_channel *channel_config;
+	uint32_t channel = dma_stm32_id_to_channel(id);
```

Comment:
Check `id < Check `id < id >= dev_config->max_channels`?`?
Ditto in `dma_stm32_stop()`.